src/main/java/cloud/prefab/client/config/UpdatingConfigResolver.java

```diff
@@ -0,0 +1,136 @@
+package cloud.prefab.client.config;
+
+import cloud.prefab.client.PrefabCloudClient;
+import cloud.prefab.domain.Prefab;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.MapDifference;
+import com.google.common.collect.Maps;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class UpdatingConfigResolver {
+
+  private final ConfigLoader configLoader;
+
+  private final PrefabCloudClient baseClient;
+  private final ConfigStoreImpl configStore;
+  private ConfigResolver configResolver;
+
+  public UpdatingConfigResolver(PrefabCloudClient baseClient, ConfigLoader configLoader) {
+    this.baseClient = baseClient;
+    this.configLoader = configLoader;
+    this.configStore = new ConfigStoreImpl();
+    this.configResolver = new ConfigResolver(configStore);
+  }
+
+  /**
+   * Return the changed config values since last update()
+   */
+  public synchronized List<ConfigChangeEvent> update() {
+    // store the old map
+    final Map<String, Optional<Prefab.ConfigValue>> before = configStore
+      .entrySet()
+      .stream()
+      .collect(
+        Collectors.toMap(
+          Map.Entry::getKey,
+          e -> configResolver.getConfigValue(e.getKey())
+        )
+      );
+
+    // load the new map
+    makeLocal();
+
+    // build the new map
+    final Map<String, Optional<Prefab.ConfigValue>> after = configStore
+      .entrySet()
+      .stream()
+      .collect(
+        Collectors.toMap(
+          Map.Entry::getKey,
+          e -> configResolver.getConfigValue(e.getKey())
+        )
+      );
+
+    MapDifference<String, Optional<Prefab.ConfigValue>> delta = Maps.difference(
+      before,
+      after
+    );
+    if (delta.areEqual()) {
+      return ImmutableList.of();
+    } else {
+      ImmutableList.Builder<ConfigChangeEvent> changeEvents = ImmutableList.builder();
+
+      // removed config values
+      delta
+        .entriesOnlyOnLeft()
+        .forEach((key, value) ->
+          changeEvents.add(new ConfigChangeEvent(key, value, Optional.empty()))
+        );
+
+      // added config values
+      delta
+        .entriesOnlyOnRight()
+        .forEach((key, value) ->
+          changeEvents.add(new ConfigChangeEvent(key, Optional.empty(), value))
+        );
+
+      // changed config values
+      delta
```

Comment:
I noticed Value subclasses don't override equals/hashcode so i think all the values there before and after would be considered changed?